test/e2e/cluster_upgrade_runtimesdk.go

```diff
@@ -228,11 +229,28 @@ func clusterUpgradeWithRuntimeSDKSpec(ctx context.Context, inputGetter func() cl
 			WaitForNodesReady: input.E2EConfig.GetIntervals(specName, "wait-nodes-ready"),
 		})
 
+		By("Deleting the workload cluster")
+		cluster := &clusterv1.Cluster{}
+		Eventually(func() error {
+			return input.BootstrapClusterProxy.GetClient().Get(ctx, client.ObjectKey{Namespace: namespace.Name, Name: clusterName}, cluster)
+		}).Should(Succeed())
+
+		// Dump all the logs from the workload cluster before deleting them.
+		input.BootstrapClusterProxy.CollectWorkloadClusterLogs(ctx, cluster.Namespace, cluster.Name, filepath.Join(input.ArtifactFolder, "clusters", cluster.Name))
+
+		framework.DeleteCluster(ctx, framework.DeleteClusterInput{
+			Deleter: input.BootstrapClusterProxy.GetClient(),
+			Cluster: cluster,
+		})
+
+		beforeClusterDeleteHandler(ctx, input.BootstrapClusterProxy.GetClient(), namespace.Name, clusterName, input.E2EConfig.GetIntervals(specName, "wait-machine-upgrade"))
```

Comment:
>Why are we using wait-machine-upgrade across handlers? 

Here it's an error - should be wait-cluster-delete. For the BeforeCreate case it's wait-cluster and then it's wait-machine-upgrade in the Upgrade handlers because that's the interval used across that test.

Reply:
But we're not actually waiting for machine upgrade or cluster delete right?

Essentially the Cluster is pretty much immediately blocked, then we unblock by changing the hook response and then we're waiting for that to be reflected.

So I wonder if we're just trying to wait long enough for the reconcile to requeue with the exponential backoff. 

We are currently returning Failure, which triggers an exponential backoff in the controller. Could it be that we should rather return retryAfterSeconds? I think this is what we actually want to test (the "blocking" not the "failure" behavior). This would also give us a relatively constant requeue based on the retryAfterSeconds value we return

Reply:
>But we're not actually waiting for machine upgrade or cluster delete right?

We are waiting for something close to it (often with a lower bar) for the blocking condition to become false i.e. unblocked - so I think we should still use intervals in that case e.g.e we are waiting for cluster delete to check if the BeforeClusterDelete hook unblocks.